### PR TITLE
Launch options

### PR DESCRIPTION
**Pull request by matemax**

### CODE REVIEW

File: examples/sdk_examples/launch_options.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+from lunavl.sdk.faceengine.engine import VLFaceEngine
+from lunavl.sdk.faceengine.setting_provider import DetectorType
+from lunavl.sdk.image_utils.image import VLImage
+from lunavl.sdk.launch_options import DeviceClass, LaunchOptions
+
+COUNT = 1000
+BATCH_SIZE = 10
+image = EXAMPLE_O
+
+faceEngine = VLFaceEngine()
```
Comment (matemax) [inline examples/sdk_examples/launch_options.py]: .

File: examples/sdk_examples/vldetection_example.py [hunk 2]
```diff
@@ -2,17 +2,27 @@
...
     Estimate all attributes
     """
-    VLFaceDetector.initialize()
-    detector = VLFaceDetector()
+    estimatorsSettings = EstimatorsSettings(
+        descriptor=FaceDescriptorEstimatorSettings(
+            descriptorVersion=59, launchOptions=LaunchOptions(deviceClass=DeviceClass.cpu)
+        )
+    )
+    detectorSettings = FaceDetectorSettings(launchOptions=LaunchOptions(deviceClass=DeviceClass.gpu))
```
Comment (matemax) [inline examples/sdk_examples/vldetection_example.py]: .

File: lunavl/sdk/async_task.py [hunk 3]
```diff
@@ -1,16 +1,25 @@
-from typing import TypeVar, Generic, Callable, Generator, List, Type, Optional, Tuple
+from abc import abstractmethod
+from typing import Any, Callable, Generator, Generic, List, Optional, Protocol, Tuple, Type, TypeVar
 
 from FaceEngine import FSDKErrorResult  # pylint: disable=E0611,E0401
 
 from lunavl.sdk.errors.exceptions import assertError
 
+
+class TResult(Protocol):
```
Comment (matemax) [inline lunavl/sdk/async_task.py:9]: .

File: lunavl/sdk/detectors/facedetector.py [hunk 4]
```diff
@@ -297,11 +298,20 @@ class FaceDetector:
...
 
-    def __init__(self, detectorPtr, detectorType: DetectorType):
+    def __init__(self, detectorPtr, detectorType: DetectorType, launchOptions: LaunchOptions):
         self._detector = detectorPtr
-        self.detectorType = detectorType
+        self._detectorType = detectorType
+        self._launchOptions = launchOptions
+
+    @property
+    def detectorType(self) -> DetectionType:
```
Comment (matemax) [inline lunavl/sdk/detectors/facedetector.py]: .

File: lunavl/sdk/estimators/base.py [hunk 5]
```diff
@@ -17,19 +18,24 @@ class BaseEstimator(ABC):
         _coreEstimator: core estimator
     """
 
-    __slots__ = ("_coreEstimator",)
+    __slots__ = ("_coreEstimator", "_launchOptions")
 
-    def __init__(self, coreEstimator):
+    def __init__(self, coreEstimator, launchOptions: LaunchOptions):
```
Comment (matemax) [inline lunavl/sdk/estimators/base.py:24]: .

File: lunavl/sdk/estimators/body_estimators/human_descriptor.py [hunk 6]
```diff
@@ -23,14 +25,19 @@ class HumanDescriptorEstimator(BaseEstimator):
     """
 
     #  pylint: disable=W0235
-    def __init__(self, coreExtractor: IDescriptorExtractorPtr, humanDescriptorFactory: "HumanDescriptorFactory"):
+    def __init__(
+        self,
+        coreExtractor: IDescriptorExtractorPtr,
+        humanDescriptorFactory: "HumanDescriptorFactory",
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/human_descriptor.py:31]: .

File: lunavl/sdk/image_utils/image.py [hunk 7]
```diff
@@ -139,6 +143,24 @@ def load(cls, colorFormat: str) -> "ColorFormat":
...
+    gpu = CoreMemoryResidence.MemoryGPU
+    # npu = CoreMemoryResidence.MemoryNPU
+
+
+class ResidenceDevice:
+    __slots__ = ("memoryResidence", "deviceId")
+
+    def __init__(self, memoryResidence: MemoryResidence = MemoryResidence.cpu, deviceId: Optional[int] = None):
+        self.memoryResidence = memoryResidence
+        self.deviceId = deviceId
```
Comment (matemax) [inline lunavl/sdk/image_utils/image.py]: .

File: lunavl/sdk/faceengine/engine.py [hunk 8]
```diff
@@ -120,25 +129,40 @@ def createFaceDetector(self, detectorType: DetectorType) -> FaceDetector:
...
+            self._faceEngine.createHeadPoseEstimator(launchOptions=launchOptions.coreLaunchOptions), launchOptions
+        )
 
-    def createWarpQualityEstimator(self) -> WarpQualityEstimator:
+    def createWarpQualityEstimator(self, launchOptions: Optional[LaunchOptions] = None) -> WarpQualityEstimator:
         """
         Create an image quality estimator
 
         Returns:
             estimator
```
Comment (matemax) [inline lunavl/sdk/faceengine/engine.py:167]: .

File: lunavl/sdk/detectors/facedetector.py [hunk 9]
```diff
@@ -198,7 +199,7 @@ def collectReDetectionsResult(
 
 def collectDetectionsResult(
     fsdkDetectRes: IFaceDetectionBatchPtr,
-    images: Union[List[Union[VLImage, ImageForDetection]], List[ImageForRedetection]],
+    images: Union[List[VLImage], List[Union[VLImage, ImageForDetection]], List[ImageForRedetection]],
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/facedetector.py:202]: the same typing fix for `_collectDetectionsResult`?

File: lunavl/sdk/detectors/facedetector.py [hunk 10]
```diff
@@ -294,14 +295,25 @@ class FaceDetector:
...
 
-    def __init__(self, detectorPtr, detectorType: DetectorType):
+    def __init__(self, detectorPtr, detectorType: DetectorType, launchOptions: LaunchOptions):
         self._detector = detectorPtr
-        self.detectorType = detectorType
+        self._detectorType = detectorType
+        self._launchOptions = launchOptions
+
+    @property
+    def detectorType(self) -> DetectionType:
```
Comment (gluhar2006) [inline lunavl/sdk/detectors/facedetector.py]:
```suggestion
    def detectorType(self) -> DetectorType:
```

File: lunavl/sdk/estimators/face_estimators/livenessv1.py [hunk 11]
```diff
@@ -140,17 +136,25 @@ class LivenessV1Estimator(BaseEstimator):
...
 
-        Args:
-            coreEstimator: core estimator
-        """
-        super().__init__(coreEstimator)
+    @overload
+    def estimate(
+        self,
+        faceDetection: FaceDetection,
+        qualityThreshold: Optional[float],
```
Comment (gluhar2006) [inline lunavl/sdk/estimators/face_estimators/livenessv1.py]:
```suggestion
        qualityThreshold: Optional[float] = None,
```

File: lunavl/sdk/faceengine/engine.py [hunk 12]
```diff
@@ -336,66 +406,103 @@ def createIndexBuilder(self, descriptorVersion: int = 0, capacity: int = 0) -> I
...
+            self._faceEngine.createBackgroundEstimator(launchOptions=launchOptions.coreLaunchOptions), launchOptions
+        )
 
-    def createImageColorTypeEstimator(self) -> ImageColorTypeEstimator:
+    def createImageColorTypeEstimator(self, launchOptions: Optional[LaunchOptions] = None) -> ImageColorTypeEstimator:
         """
         Create a image color type estimator.
+
+        Args:
+            launchOptions: estimator launch options
```
Comment (gluhar2006) [inline lunavl/sdk/faceengine/engine.py:572]: maybe it's better to remove the `Args` section than to add it to all the other `createSomething`?

File: lunavl/sdk/launch_options.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class DeviceClass(Enum):
+    """
+    Device enum
+    """
+
+    cpu = "cpu"
+    gpu = "gpu"
+    # npu = "npu"
```
Comment (gluhar2006) [inline lunavl/sdk/launch_options.py]:
```suggestion
```

File: lunavl/sdk/launch_options.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        deviceClass: Optional[DeviceClass] = None,
+        deviceId: Optional[int] = None,
+        runConcurrently: Optional[bool] = None,
+    ):
+        self._coreLaunchOptions = CoreFE.LaunchOptions()
+        if deviceClass:
+            if deviceClass == DeviceClass.gpu:
+                device = CoreFE.DeviceClass.GPU
+            # elif deviceClass == DeviceClass.npu:
+            #     device = CoreFE.DeviceClass.NPU
```
Comment (gluhar2006) [inline lunavl/sdk/launch_options.py]:
```suggestion
```

File: lunavl/sdk/detectors/facedetector.py (hunk 9 above)
Comment (matemax) [reply lunavl/sdk/detectors/facedetector.py:202]: mypy changed his mind

File: lunavl/sdk/launch_options.py (hunk 13 above)
Comment (matemax) [reply lunavl/sdk/launch_options.py]: ++

File: lunavl/sdk/launch_options.py (hunk 14 above)
Comment (matemax) [reply lunavl/sdk/launch_options.py]: ++

File: lunavl/sdk/faceengine/engine.py (hunk 12 above)
Comment (matemax) [reply lunavl/sdk/faceengine/engine.py:572]: Add to all, it's not difficult for me

**Pull request merged**